runtime/onert/core/src/compiler/ExecutorFactory.cc

```diff
@@ -721,17 +721,14 @@ exec::IExecutor *ExecutorFactory::createTrainableExecutor(
     tdata.optim_info = training_info.optimizerInfo();
 
     // TODO Remove dynamic_cast
-    try
-    {
-      const auto backend = pair.first;
-      const auto tbackend = dynamic_cast<const backend::train::ITrainableBackend *>(backend);
-      tbackend_contexts.emplace(backend, tbackend->newContext(std::move(tdata)));
-    }
-    catch (const std::bad_cast &)
+    const auto backend = pair.first;
+    const auto tbackend = dynamic_cast<const backend::train::ITrainableBackend *>(backend);
+    if (!tbackend)
     {
       throw std::runtime_error("ExecutorFactory: Invalid backend - TrainableExecutor does not "
                                "support non-trainble backends");
     }
+    tbackend_contexts.emplace(backend, tbackend->newContext(std::move(tdata)));
```

Comment:
I think this doesn't really prevent future committers from making the same mistake.

What if we instead introduce a custom cast function which does `dynamic_cast` + `nullptr check`?

Reply:
That's really good option. However, we should think further. If we introduce a custom cast function, there are many combinations of handling dynamic_cast. For example,
- dynamic_cast + null pointer
    - throw exception
    - ignore(continue to process forward)
- dynamic_cast + reference
    - throw exception
    - ignore

Simply, we may introduce 4 types of custom functions. I think it could cost expensive.